sel/src/Sel/SecretKey/EncryptedStream.hs

```diff
@@ -0,0 +1,542 @@
+    , Show
+      -- ^ @since 0.0.1.0
+    , Ord
+      -- ^ @since 0.0.1.0
+    )
+  deriving
+    ( Display
+      -- ^ @since 0.0.1.0
+    )
+    via (ShowInstance EncryptedStreamError)
+
+-- == Encryption ==
+
+-- | Initialise a stream to which you will push encrypted message.
+-- This function returns the 'Header' and the 'SecretKey' that are needed for the peer to
+-- decrypt the stream.
+--
+-- @since 0.0.1.0
+initPushStream
+  :: Multipart s
+  -- ^ Cryptographic state
+  -> IO (Either EncryptedStreamError (Header, SecretKey))
+initPushStream (Multipart statePtr) = do
+  headerForeignPtr <- liftIO $ Foreign.mallocForeignPtrBytes (fromIntegral cryptoSecretStreamXChaCha20Poly1305HeaderBytes)
+  (SecretKey secretKeyForeignPtr) <- newSecretKey
+  Foreign.withForeignPtr headerForeignPtr $ \headerPtr ->
+    Foreign.withForeignPtr secretKeyForeignPtr $ \keyPtr -> do
+      result <-
+        cryptoSecretStreamXChaCha20Poly1305InitPush
+          statePtr
+          headerPtr
+          keyPtr
+      case result of
+        0 -> pure $ Right (Header headerForeignPtr, SecretKey secretKeyForeignPtr)
+        _ -> pure $ Left InputStreamInitError
+
+-- |
+--
+-- @since 0.0.1.0
+pushToStream
+  :: Multipart s
+  -- ^ The cryptographic state
+  -> StrictByteString
+  -- ^ The message to encrypt
+  -> Maybe StrictByteString
+  -- ^ Additional, optional data that can be included.
+  -> StreamTag
+  -- ^ Tag that accompanies the resulting 'CipherText'.
+  -> IO (Either EncryptedStreamError CipherText)
+pushToStream (Multipart statePtr) message mData tag =
+  case mData of
+    Just additionalData -> do
+      let additionalDataLength = BS.length additionalData
+      Foreign.allocaBytes additionalDataLength $ \additionalDataPtr ->
+        doPushToStream additionalDataPtr (fromIntegral additionalDataLength)
+    Nothing -> doPushToStream Foreign.nullPtr 0
+  where
+    doPushToStream :: Ptr CUChar -> CULLong -> IO (Either EncryptedStreamError CipherText)
+    doPushToStream additionalDataPointer additionalDataLength =
+      BS.unsafeUseAsCStringLen message $ \(cString, cStringLen) -> do
+        let cipherTextLength =
+              fromIntegral cryptoSecretStreamXChaCha20Poly1305ABytes + cStringLen
+        cipherTextFPtr <- Foreign.mallocForeignPtrBytes cipherTextLength
+        Foreign.allocaBytes cipherTextLength $ \cipherTextBuffer -> do
+          result <-
+            cryptoSecretStreamXChaCha20Poly1305Push
+              statePtr
+              cipherTextBuffer
+              Foreign.nullPtr
+              (Foreign.castPtr @CChar @CUChar cString)
+              (fromIntegral @Int @CULLong cStringLen)
+              additionalDataPointer
+              additionalDataLength
+              (streamTagToConstant tag)
+          case result of
+            0 -> do
+              Foreign.withForeignPtr cipherTextFPtr $ \cipherTextHomePtr ->
+                memcpy cipherTextBuffer cipherTextHomePtr (fromIntegral cipherTextLength)
```

Comment:
This is a bit inefficient: you allocate a temporary buffer, only to then copy it immediately on success.